src/biogeophys/HillslopeHydrologyMod.F90

```diff
@@ -0,0 +1,1140 @@
+       waterfluxbulk_inst,wateratm2lndbulk_inst)
+    !
+    ! !DESCRIPTION:
+    ! Calculate discharge from stream channel
+    !
+    ! !USES:
+    use LandunitType    , only : lun                
+    use GridcellType    , only : grc                
+    use ColumnType      , only : col                
+    use WaterFluxBulkType   , only : waterfluxbulk_type
+    use WaterStateBulkType  , only : waterstatebulk_type
+    use Wateratm2lndBulkType, only : wateratm2lndbulk_type
+    use spmdMod         , only : masterproc
+    use clm_varcon      , only : spval, ispval, grlnd 
+    use landunit_varcon , only : istsoil
+    use clm_time_manager, only : get_step_size_real
+
+    !
+    ! !ARGUMENTS:
+    type(bounds_type), intent(in) :: bounds
+    type(waterstatebulk_type), intent(inout) :: waterstatebulk_inst
+    type(waterfluxbulk_type),  intent(inout) :: waterfluxbulk_inst
+    type(wateratm2lndbulk_type), intent(inout) :: wateratm2lndbulk_inst
+
+    integer               :: c, l, g, i, j
+    real(r8) :: qflx_surf_vol                           ! volumetric surface runoff (m3/s)
+    real(r8) :: qflx_drain_perched_vol                  ! volumetric perched water table runoff (m3/s)
+    real(r8) :: qflx_rsub_sat_vol                       ! volumetric correction runoff (m3/s)
+    real(r8) :: dtime                                   ! land model time step (sec)
+
+    character(len=*), parameter :: subname = 'HillslopeUpdateStreamWater'
+
+    !-----------------------------------------------------------------------
+    associate(                                                            & 
+         stream_water_volume     =>    waterstatebulk_inst%stream_water_lun    , & ! Input/Output:  [real(r8) (:)   ] stream water volume (m3)
+         qstreamflow             =>    waterfluxbulk_inst%qstreamflow_lun      , & ! Input:  [real(r8) (:)   ] stream water discharge (m3/s)
+         qdischarge              =>    waterfluxbulk_inst%qdischarge_col       , & ! Input: [real(r8) (:)   ]  discharge from columns (m3/s)
+         qflx_drain_perched      =>    waterfluxbulk_inst%qflx_drain_perched_col, &! Input:  [real(r8) (:)   ]  column level sub-surface runoff (mm H2O /s)
+         qflx_rsub_sat           =>    waterfluxbulk_inst%qflx_rsub_sat_col    , & ! Input:  [real(r8) (:)   ]  column level correction runoff (mm H2O /s)
+         qflx_surf               =>    waterfluxbulk_inst%qflx_surf_col        , & ! Input: [real(r8) (:)   ]  total surface runoff (mm H2O /s)
+         stream_water_depth      =>    waterstatebulk_inst%stream_water_depth_lun & ! Output:  [real(r8) (:)   ] stream water depth (m)
+         )
+
+       ! Get time step
+       dtime = get_step_size_real()
+
+       do l = bounds%begl,bounds%endl
+          if(lun%itype(l) == istsoil) then
+             g = lun%gridcell(l)
+             
+             do c = lun%coli(l), lun%colf(l)
+                if (col%is_hillslope_column(c)) then            
```

Comment:
- [ ] This should either check col%active(c) or be reworked to use a filter (which would only contain active points) – otherwise you can be trying to do calculations on garbage values.